### DISCUSSION TITLE
Not correct convertation svg->png using php code

### DISCUSSION THREAD

**Discussion by fedushok** [ImageMagick/ImageMagick#5309] [Development]
svg file [(m.svg)](https://ucarecdn.com/bc3d17b0-9a83-406f-bab5-d40704f0d075/)

I got some variants of svg->png conversation
**First variant of png** (right version):
<a href="https://ibb.co/CvZjRnL"><img src="https://i.ibb.co/RPmXRC8/m1.png" alt="m1" border="0"></a>
**Second variant of png**:
<a href="https://ibb.co/ZMRWvqt"><img src="https://i.ibb.co/pL6JNYG/m2.png" alt="m2" border="0"></a>
**Third variant of png**:
<a href="https://ibb.co/vqsNjzD"><img src="https://i.ibb.co/54YQRc2/m3.png" alt="m3" border="0"></a>

First variant of png was made by convert command.
Second variant was made by php code.

_My programs:_
Ubuntu 22.04
php 8.1.7
Inkscape 1.1.2 (0a00cf5339, 2022-02-04) libxml2 librsvg2-dev (sudo apt
install ... )
ImageMagick6-6.9.11-60 ( ./configure --with-png=yes --with-rsvg=yes && make && sudo make
install && sudo ldconfig /usr/local/lib) (in phpinfo it is said that php-imagick 3.7.0 was compiled with this
version of IM)
php-imagick 3.7.0 (sudo apt install ... )

_Information given by some commands:_

identify -list format | grep SVG ->
     MSVG* rw+   ImageMagick's own SVG internal renderer
      SVG* rw+   Scalable Vector Graphics (RSVG 2.52.5)
     SVGZ* rw+   Compressed Scalable Vector Graphics (RSVG 2.52.5)

phpinfo -> ImageMagick supported formats: … , MSVG, ... SVG, SVGZ, ...

convert -list configure ->
...
DELEGATES      cairo fontconfig freetype jbig jng jpeg lzma png rsvg tiff x xml zlib
...

identify --version ->
Version: ImageMagick 6.9.11-60 Q16 x86_64 2021-01-24 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC OpenMP(4.5)
Delegates (built-in): cairo fontconfig freetype jbig jng jpeg lzma png rsvg tiff x xml zlib

_Results given by some commands:_

rsvg-convert -o m1.png m.svg -> **second variant of png**
convert m.svg m.png -> **first variant of png**
convert SVG:m.svg m.png -> **first variant of png**
convert MSVG:m.svg m.png -> **third variant of png**

rsvg-convert gives the second variant of png (not correct). I tested another svg files (with another mistakes) and got the same result with the result made by php code.
All my svg are rendered correctly in the Inkscape and various browsers.

If compile IM without --with-png=yes option, then php code and convert command gives **third variant of png**.

My questions are:
1. How to make php-imagick - imagemagick convert svg->png correctly?
2. How could be explained the difference in svg->png result between
convert command and php-imagick? As i understand convert command and
php-imagick use the same imagemagick, but the results are different.
3. How can I know which delegate IM uses by default?

**Comment by snibgo** [comment]
MSVG is very simple, and doesn't process the more advanced SVG features. Using that, with IM v7.1.0-32, I get:
```
magick msvg:m.svg x.png
```
[img x.png]
![x](https://user-images.githubusercontent.com/33812368/178121831-9d52aac3-1e8c-46ec-8809-d3b0e2fc3630.png)

When Inkscape is on my system path:
```
magick m.svg xi.png
```
[img xi.png]

![xi](https://user-images.githubusercontent.com/33812368/178121838-4231c9f2-e545-476c-b887-69c4e5d5ca7e.png)

With rsvg-convert v2.40.20:
```
rsvg-convert -o r.png m.svg
```
[img r.png]

![r](https://user-images.githubusercontent.com/33812368/178121843-5682ffb9-18c9-4a53-bfd8-59b1ed1b053b.png)

I have no experience with PHP or rsvg-convert. 

IM will use Inkscape if it is on the system path. I think PHP has a different system path to the command line.


> 3. How can I know which delegate IM uses by default?

Include "-verbose" in your command. Current v7 IM will tell you if it is using RSVG, MSVG or Inkscape.

**Comment by fedushok** [comment]
Here is info about $PATH
```
env ->
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin:/snap/bin
phpinfo ->
Apache Environment PATH
     /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
PHP Variables $_SERVER['PATH'] | /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin
```
They are quiet the same. Except "games".
```
which inkscape ->
/usr/bin/inkscape
```
convert -verbose says that Inkscape is a delegate. 
Is it possible to know which delegate is used in php code? May be IM has log file or something like that.
